File: component-acgnews/src/main/java/com/rabtman/acgnews/mvp/presenter/ZeroFiveNewsItemPresenter.java
package com.rabtman.acgnews.mvp.presenter;

import com.rabtman.acgnews.base.constant.HtmlConstant;
import com.rabtman.acgnews.mvp.contract.ZeroFiveNewsContract;
import com.rabtman.acgnews.mvp.model.jsoup.ZeroFiveNewsPage;
import com.rabtman.common.base.CommonSubscriber;
import com.rabtman.common.base.mvp.BasePresenter;
import com.rabtman.common.di.scope.FragmentScope;
import com.rabtman.common.utils.RxUtil;
import javax.inject.Inject;

/**
 * @author Rabtman
 */
@FragmentScope
public class ZeroFiveNewsItemPresenter extends
    BasePresenter<ZeroFiveNewsContract.Model, ZeroFiveNewsContract.View> {

  //当前页面位置
  private int pageNo = 1;

  @Inject
  public ZeroFiveNewsItemPresenter(ZeroFiveNewsContract.Model model,
      ZeroFiveNewsContract.View rootView) {
    super(model, rootView);
  }

  public void getAcgNewsList() {
    pageNo = 1;
    addSubscribe(
        mModel.getAcgNews(HtmlConstant.ZERO_FIVE_URL + mView.getNewsUrl(pageNo))
            .compose(RxUtil.<ZeroFiveNewsPage>rxSchedulerHelper())
            .subscribeWith(new CommonSubscriber<ZeroFiveNewsPage>(mView) {

              @Override
              public void onError(Throwable e) {
                super.onError(e);
                mView.showPageError();
              }

              @Override
              public void onComplete() {
                mView.hideLoading();
              }

              @Override
              public void onNext(ZeroFiveNewsPage zeroFiveNewsPage) {
                if (zeroFiveNewsPage.getZeroFiveNewsList().size() > 0) {
                  mView.showAcgNews(zeroFiveNewsPage.getZeroFiveNewsList());
                  mView.showPageContent();
                } else {
                  mView.showPageEmpty();
                }
              }
            })
    );
  }

  public void getMoreAcgNewsList() {
    addSubscribe(
        mModel.getAcgNews(HtmlConstant.ZERO_FIVE_URL + mView.getNewsUrl(++pageNo))
            .compose(RxUtil.<ZeroFiveNewsPage>rxSchedulerHelper())
            .subscribeWith(new CommonSubscriber<ZeroFiveNewsPage>(mView) {
              @Override
              public void onNext(ZeroFiveNewsPage zeroFiveNewsPage) {
                int pageCount;
                try {
                  pageCount = Integer.parseInt(zeroFiveNewsPage.getPageCount());
                } catch (NumberFormatException e) {
                  e.printStackTrace();
                  pageCount = 1;
                }
                mView.showMoreAcgNews(zeroFiveNewsPage.getZeroFiveNewsList(),
                    pageNo < pageCount);
              }

              @Override
              public void onError(Throwable e) {
                super.onError(e);
                mView.onLoadMoreFail();
                pageNo--;
              }
            })
    );
  }

}

File: component-acgnews/src/main/java/com/rabtman/acgnews/mvp/presenter/ZeroFiveNewsItemPresenter.kt
package com.rabtman.acgnews.mvp.presenter

import com.rabtman.acgnews.base.constant.HtmlConstant
import com.rabtman.acgnews.mvp.contract.ZeroFiveNewsContract
import com.rabtman.acgnews.mvp.model.jsoup.ZeroFiveNewsPage
import com.rabtman.business.base.CommonSubscriber
import com.rabtman.common.base.mvp.BasePresenter
import com.rabtman.common.di.scope.FragmentScope
import com.rabtman.common.utils.RxUtil
import javax.inject.Inject

/**
 * @author Rabtman
 */
@FragmentScope
class ZeroFiveNewsItemPresenter @Inject constructor(model: ZeroFiveNewsContract.Model,
                                                    rootView: ZeroFiveNewsContract.View) : BasePresenter<ZeroFiveNewsContract.Model, ZeroFiveNewsContract.View>(model, rootView) {
    //当前页面位置
    private var pageNo = 1
    fun getAcgNewsList() {
        pageNo = 1
        addSubscribe(
                mModel.getAcgNews(HtmlConstant.ZERO_FIVE_URL + mView.getNewsUrl(pageNo))
                        .compose(RxUtil.rxSchedulerHelper())
                        .subscribeWith(object : CommonSubscriber<ZeroFiveNewsPage>(mView) {
                            override fun onError(e: Throwable) {
                                super.onError(e)
                                mView.showPageError()
                            }

                            override fun onComplete() {
                                mView.hideLoading()
                            }

                            override fun onNext(zeroFiveNewsPage: ZeroFiveNewsPage) {
                                zeroFiveNewsPage.zeroFiveNewsList?.takeIf { it.isNotEmpty() }?.let {
                                    mView.showAcgNews(it)
                                    mView.showPageContent()
                                } ?: run {
                                    mView.showPageEmpty()
                                }
                            }
                        })
        )
    }

    fun getMoreAcgNewsList() {
        addSubscribe(
                mModel.getAcgNews(HtmlConstant.ZERO_FIVE_URL + mView.getNewsUrl(++pageNo))
                        .compose(RxUtil.rxSchedulerHelper())
                        .subscribeWith(object : CommonSubscriber<ZeroFiveNewsPage>(mView) {
                            override fun onNext(zeroFiveNewsPage: ZeroFiveNewsPage) {
                                mView.showMoreAcgNews(zeroFiveNewsPage.zeroFiveNewsList,
                                        pageNo < zeroFiveNewsPage.getPageCount())
                            }

                            override fun onError(e: Throwable) {
                                super.onError(e)
                                mView.onLoadMoreFail()
                                pageNo--
                            }
                        })
        )
    }
}